python/mxnet/io/im2rec.py

```diff
@@ -0,0 +1,270 @@
+    if color not in [1, 0, -1]:
+        raise ValueError("Invalid value for color parameter. "
+                         + "Should be 1, 0 or -1")
+    if encoding not in ['.jpg', '.png']:
+        raise ValueError("Encoding should be either .jpg or .png")
+    if not isinstance(quality, int):
+        raise ValueError("Quality should be an integer value. "
+                         + "range for quality of JPEG encoding is 1-100, "
+                         + "and range for PNG compression for encoding is 1-9")
+    if not pack_labels:
+        logging.info("pack_labels is set as False. We will create a copy of the "
+                     + "list file to align with the created rec file.")
+
+def im2rec(list_file, output_path, transformations=None, num_workers=mp.cpu_count() - 1,
+           batch_size=4096, pack_labels=True, color=1, encoding='.jpg',
+           quality=95, pass_through=False, error_limit=0):
+    """
+    API to convert the input image dataset into recordIO file format.
+
+    Parameters
+    ----------
+    list_file : str
+        List file name
+    output_path : str
+        output file path
+    transformations : Transforms object
+        Transforms object
+    num_workers : int. default (number of cpu cores - 1)
+        number of workers, ideally do not set this value to a number
+        greater than the number of cpu cores in your machine.
+    batch_size : int, default 4096
+        number of images to process in one batch.
+    pack_labels : bool, default True
+        boolean to determine if multi-dimensional labels should be packed.
+    color : int, default 1
+        specify the color mode of the loaded image.
+        1: Loads a color image
+        0: Loads image in grayscale mode.
+        -1: Loads image unchanged, including alpha channel.
+    encoding : str, default '.jpg'
+        type of encoding for the images.
+        it can be one of ['.jpg', '.png']
+    quality : int, default 95
+        JPEG quality for encoding, the range is 1-100;
+        or PNG compression for encoding, the range is 1-9
+    pass_through : bool, default False
+        should the image be packed without any transformation.
+    error_limit: int, default 0
+        While creating the dataset there might be errors due
+        to corrupt images and inability to pack them in recordIO
+        format. This parameter decides how many such errors can
+        be swallowed while generating the rec file.
+
+    Returns
+    -------
+    output_path : str
+        output path of the rec file
+
+    """
+    _validate_params(list_file, output_path, pack_labels, color,
+                     encoding, quality, pass_through)
+    fname = os.path.basename(list_file).split('.')[0]
+    shared_obj_mgr = SharedObjectManager()
+    shared_obj_mgr.start()
+
+    data_batch_iter = _read_list(list_file, batch_size)
+    # A process-safe PriorityQueue
+    out_q = shared_obj_mgr.PriorityQueue(batch_size)
```

Comment:
I am not sure why you want a priority queue here?

Reply:
we need a priority queue to ensure that the create recio file has the images in the same order as list in the .lst file. we preferably do not want to alter the lst file and we need to have the lst file and rec file in sync.